### PR TITLE
[autoscaler] Command runner interactivity

### LINKED ISSUE #8704: IOCTL error in Autoscaler
### What is the problem?
The following error message is printed repeatedly when using the autoscaler. It appears many times during `ray up ...`
```
bash: cannot set terminal process group (-1): Inappropriate ioctl for device
bash: no job control in this shell
```

*Ray version and other system information (Python version, TensorFlow version, OS):*

### Reproduction (REQUIRED)
Please provide a script that can be run to reproduce the issue. The script should have **no external library dependencies** (i.e., use fake or mock data / environments):

If we cannot run your script, we cannot fix your issue.

- [x] I have verified my script runs in a clean environment and reproduces the issue.
- [ ] I have verified the issue also occurs with the [latest wheels](https://docs.ray.io/en/latest/installation.html).

### PR DESCRIPTION
**Pull request by wuisawesome**
## Why are these changes needed?

The autoscaler runs all of its commands in interactive mode with `bash --login -c -i 'true <command>`. This expects a tty to be allocated, so it doesn't make sense to have it as a separate option (this causes ioctl errors to be printed). Instead, we should simply have an interactive option which runs commands interactively and allocates a tty at the same time.

## Related issue number

Closes #8704

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://ray.readthedocs.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.

### CODE REVIEW

**Comment by richardliaw** [comment]
awesome - does this work on AWS autoscaler?

**Comment by wuisawesome** [comment]
> awesome - does this work on AWS autoscaler?

It does for me at least  😛

**Comment by richardliaw** [comment]
Merge conflict - overall changes look ok to me, I'll try it out soon.

**Review by richardliaw — COMMENTED** [review]
`Shared connection to 3.95.19.139 closed.`

This keeps showing up on each line now after each command is executed. Is it possible to avoid this?

**Comment by ijrsvt** [comment]
@richardliaw It seems like that error is printed if a tty is allocated. I think that is probably okay because that is printed for `exec`, `submit` and `dashboard` (and possibly a few more) because those commands already use `allocate_tty`. 

Personally, I think `shared connection closed` is far less concerning than the `ioctl` and `job control` errors

**Comment by ijrsvt** [comment]
Per each old command in startup the commands would look like:
```
2020-06-01 11:35:01,669	INFO updater.py:257 -- NodeUpdater: HEAD: Running uptime on HEAD...
bash: cannot set terminal process group (-1): Inappropriate ioctl for device
bash: no job control in this shell
 18:35:02 up 4 days,  1:14,  6 users,  load average: 0.10, 0.15, 0.11
2020-06-01 11:35:02,299	INFO log_timer.py:17 -- NodeUpdater: HEAD: Got remote shell [LogTimer=631ms]
```

Whereas the new output would be
```
2020-06-01 11:32:08,140	INFO updater.py:257 -- NodeUpdater: HEAD: Running uptime on HEAD...
 18:32:08 up 4 days,  1:11,  7 users,  load average: 0.22, 0.13, 0.10
Shared connection to HEAD closed.
2020-06-01 11:32:08,820	INFO log_timer.py:17 -- NodeUpdater: HEAD: Got remote shell [LogTimer=681ms]
```

**Comment by wuisawesome** [comment]
I think this post sums it up, but this is really a quirk of SSH. 
https://unix.stackexchange.com/questions/191265/avoid-shared-connection-to-host-closed-messages

@ijrsvt Perhaps it wouldn't actually hurt to just silence `stderr`. More serious SSH errors should cause a non-zero return code which we could catch instead right?

**Comment by richardliaw** [comment]
Let's not hide stderr messages (because I don't know what the consequences of that would be). I think leaving the 'Shared connection closed' is fine. It would be nice to be able to catch and hide that specific string, but that's a nit.

**Comment by ijrsvt** [comment]
We could add something ugly like ` 2> >(grep -v "Shared Connection")` to these commands.

**Comment by wuisawesome** [comment]
@ijrsvt @richardliaw updated, let me know what y'all think

File: python/ray/autoscaler/updater.py [hunk 1]
```diff
@@ -255,14 +257,25 @@ def run(self,
...
             if with_output:
-                return self.process_runner.check_output(final_cmd)
+                if interactive:
+                    ssh_process = self.process_runner.Popen(
+                        final_cmd, stderr=subprocess.PIPE)
+                    err = self.process_runner.check_output(
+                        ["grep", "-v", "Shared Connection to .* closed"],
+                        stdin=ssh_process.stderr)
+                    if err:
+                        print(err, file=sys.stderr)
```
Comment (ericl) [inline python/ray/autoscaler/updater.py:275]: This is super hacky. Do we really need this?

File: python/ray/autoscaler/commands.py [hunk 2]
```diff
@@ -453,7 +453,7 @@ def _exec(updater, cmd, screen, tmux, port_forward=None, with_output=False):
             cmd = " ".join(cmd)
     return updater.cmd_runner.run(
         cmd,
-        allocate_tty=True,
+        interactive=bool(cmd),
```
Comment (ericl) [inline python/ray/autoscaler/commands.py:456]: Isn't this always true for non-empty commands?

File: python/ray/autoscaler/updater.py (hunk 1 above)
Comment (wuisawesome) [reply python/ray/autoscaler/updater.py:275]: Filtering out certain error messages feels inherently hacky. 

I don't have a strong opinion about this, so I'll let you you and @richardliaw decide?

Comment (richardliaw) [reply python/ray/autoscaler/updater.py:275]: If there's no cleaner solution, let's remove it

File: python/ray/autoscaler/commands.py (hunk 2 above)
Comment (wuisawesome) [reply python/ray/autoscaler/commands.py:456]: cmd can be `None`, for example if we're only doing port forwarding.

Comment (ericl) [reply python/ray/autoscaler/commands.py:456]: Is there any harm in always enabling interactive mode and just removing the flag?

Comment (wuisawesome) [reply python/ray/autoscaler/commands.py:456]: You tell me :p (I think you originally introduced the `allocate_tty` flag a few years ago).

There's technically some overhead, but I don't think it's noticeable

**Review by ericl — APPROVED** [review]

**Review by richardliaw — APPROVED** [review]

**Pull request merged**